feat: Alpha Support for Multi-remote Browser

dprevost-LMI opened the pull request:
Fixes partially https://github.com/webdriverio/expect-webdriverio/issues/106

PR serving as a starting point to support the multi-remote case.
- The goal is to have a strong groundwork and not to implement all matchers in a single PR.
- The first iteration is to implement the `toHaveTitle` browser matchers
- Element matchers will come next
- `waitUntil` was reviewed to support multiple instances better so that we can retry only those that failed, optimizing the fetching process.
   - Also, it now returns a `CompareResult` to be more flexible and returns more information for the assertions.
- The mechanism to fetch actual data and to format the error message was streamlined using an array of instances (even for non-multi-remote cases) to have a single algorithm used, but all extensible.
- Until the work is proven enough, it is kept under the **UNSTABLE** umbrella, so it is to be used at your own risk.

Depnd on:
- For per-browser value, we need https://github.com/webdriverio/webdriverio/pull/15459#issuecomment-5160734487

Draft TODOs:
- [x] Code review
- [x] Add unit tests
- [ ] Remove array expected support?

dprevost-LMI committed:
Draft first working multi-remote case

- Simple working case expecting one value

dprevost-LMI committed:
useful command to check everything (to revert)

dprevost-LMI committed:
Working multiple targets and failures

dprevost-LMI committed:
Working with multiple provided values

dprevost-LMI committed:
Add comment on future plan

dprevost-LMI committed:
Working case of optimize waitUntil running per instance + failure msg

dprevost-LMI committed:
working consolidation of waitUntil + formatFailureMessage

- Use `waitUntilResult` optimize to rerun per instance and now returns results with multiple information easing the processing. Also support condition returning an array of result or an array of promise
- Use `formatFailureMessage` both compatible for multi-remote and not
- Streamline and unify the processing of browser + get of actual value to use same algo to support easier both multi-remote and not multi-remote

dprevost-LMI committed:
Add doc for supported case for alpha version + some future plan

dprevost-LMI commented on src/utils.ts:
```diff
@@ -1,93 +1,178 @@
+): Promise<{ pass: boolean, results: CompareResult<A, E>[] }> => {
+    /**
+     * Using array algorithm to handle both single and multiple conditions uniformly
+     * Tehcnically this is an o2(n) operation but pratically, we process either a single promise with Array or an Array of promises
```
```suggestion
     * Technically, this is an o(n3) operation, but practically, we process either a single promise with Array or an Array of promises. Review later if we can simplify and only have an array of promises
```

dprevost-LMI commented on src/utils.ts:
```diff
@@ -1,93 +1,178 @@
+        try {
+            const pendingConditions = allConditionsResults.filter(({ results }) => !results.every((result) => result.result))
+
+            // TODO dprevost: verify how to handle errors for each condition
```
```suggestion
            // TODO multi-remote support: handle errors per remote more gracefully, so we report failures and throws if all remotes are in errors (and therefore still throw when not multi-remote)
```

dprevost-LMI commented on types/expect-webdriverio.d.ts:
```diff
@@ -495,20 +656,36 @@ declare namespace ExpectWebdriverIO {
+    //     verb?: string
+    //     expectation?: string
+    // }
+
```
```suggestion
```

dprevost-LMI committed:
code review

dprevost-LMI committed:
Add multi-remote alternative doc with Parametrized Tests

dprevost-LMI committed:
Add Direct Instance Access alternative + fix english grammar

dprevost-LMI committed:
Code review

dprevost-LMI committed:
Add multi-remote typing support

dprevost-LMI committed:
Add missing unit tests for waitUntilResult

dprevost-LMI committed:
fix linting

dprevost-LMI committed:
add unit test for formatMessage

dprevost-LMI committed:
Add UT for multiRemoteUtil

dprevost-LMI committed:
Code review

dprevost-LMI committed:
fix rebranching

dprevost-LMI committed:
Revert undesired formating changes

dprevost-LMI committed:
revert undesired changes

dprevost-LMI committed:
Code review

dprevost-LMI committed:
revert more undesired changes

dprevost-LMI committed:
Final code review

christian-bromann commented on docs/MultiRemote.md:
````diff
@@ -0,0 +1,155 @@
+To assert all remotes with a default value, overriding specific ones:
+```ts
+    await expect(multiRemoteBrowser).toHaveTitle({
+        default : 'My Default Site Title',
````
What does `default` mean here?

dprevost-LMI replied:
It would provide a default value to assert on all remotes. So instead of enumerating them all when you only have one exception, you can still offer one value for all and another for the exception.

christian-bromann replied:
What happens if someone calls their browser `default`? Also are there many occurrences where you have to provide an exception for a single browser? How about we settle on:

```ts
await expect(multiRemoteBrowser).toHaveTitle("value for all")
// or
await expect(multiRemoteBrowser).toHaveTitle({
  browserA: "value A",
  browserB: "value B"
})
```

And see what the community responds.

dprevost-LMI replied:
What happens if someone calls their browser default?
Yep, that was a concern I had, still thinking about it.

Also, are there many occurrences where you have to provide an exception for a single browser? 
Excellent question. The idea was to provide "flexibility" 

How about we settle on
I like it, baby steps and more incremental features, I agree.
However, does your proposal also scope out the array approach?

dprevost-LMI committed:
Fix rebase

dprevost-LMI committed:
Add missing typing test for toBeDisplayed

dprevost-LMI committed:
Add doc and future task for multi-remote

dprevost-LMI committed:
Apply suggestions from code review

dprevost-LMI committed:
fix: backport isNot ambiguous failure message from PR #1987

Backporting fixes of issue https://github.com/webdriverio/expect-webdriverio/pull/1986 and PR https://github.com/webdriverio/expect-webdriverio/pull/1987 to multi-remote

dprevost-LMI committed:
fix: Stop altering arg variable  + more robust and clear concatenation
